### PR TITLE
Convert `patterns` package to TypeScript

### LINKED ISSUE #12178: TypeScript: Convert `patterns` package

### PR DESCRIPTION
**Pull request by barklund**
## Context

What is says on the tin.

NB: This depends on the `i18n` package, which hasn't been converted to TS yet, so it has some errors in a single file due to that.

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12178

### CODE REVIEW

File: packages/patterns/src/getOpaquePattern.ts
Comment (barklund) [inline packages/patterns/src/getOpaquePattern.ts:43]: I made a slight improvement here - this is way faster than re-creating objects over and over.

File: packages/patterns/src/types.ts
Comment (miina) [inline packages/patterns/src/types.ts]: Curious why use types and not interfaces for the objects? As far as I've read so far the general recommendation seems to be using types for functions and interfaces for objects, however, it looks like with the latest TS versions, these two have become more similar. Is there any reason why you prefer types to interfaces here?

File: packages/patterns/src/getOpaquePattern.ts
Comment (miina) [inline packages/patterns/src/getOpaquePattern.ts:39]: **💡 thought:**
Should we try to use the same convention across the files for consistency, .e.g always either `Array<string>` or `string[]`? Or no preference? Not sure if this is usually considered important in TS projects. Thoughts?

File: packages/patterns/package.json
Comment (miina) [inline packages/patterns/package.json]:
```suggestion
  "source": "src/index.ts",
```
Same above as well for `src/index.js` reference.

**Review by miina — APPROVED** [review]
🔨  (LGTM pending i18n / test failures)

**Comment by miina** [comment]
Looks like there are some test failures now.

File: packages/patterns/src/getOpaquePattern.ts
Comment (swissspidy) [reply packages/patterns/src/getOpaquePattern.ts:39]: No strong preference, but we could add a lint rule for this: https://typescript-eslint.io/rules/array-type
I think we've mostly used `string[]` so far, also in the PHP code base

File: packages/patterns/package.json
Comment (barklund) [reply packages/patterns/package.json]: Good catch!

File: packages/patterns/src/getOpaquePattern.ts
Comment (barklund) [reply packages/patterns/src/getOpaquePattern.ts:39]: Yeah, let's go for the more compact syntax.

File: packages/patterns/src/types.ts
Comment (barklund) [reply packages/patterns/src/types.ts]: They're mostly interchangeable and I don't think there's any real difference in these cases. Interfaces are easier to extend, but this doesn't require extension ever, so not require. The union type though can only be a type (you can't create a union interface IIRC), so the final thing here has to be type, but the other things could be interfaces.

Comment (barklund) [reply packages/patterns/src/types.ts]: I did improve the code around this a bit though, creating a `Gradient` intermediate type. Not sure if this is best-practice, but it works out pretty well.

File: packages/patterns/src/types.ts
Comment (barklund) [inline packages/patterns/src/types.ts:66]: This is a bit annoying by the way. Yes, it makes solid patterns more compact, because the solid type is implicit, but it makes type narrowing more complex, as we can't just get auto-narrowing by validating the type, but we also have to check for lack of type and then manually cast. But I still think it's worth it.

File: packages/patterns/tsconfig.json
Comment (swissspidy) [inline packages/patterns/tsconfig.json:7]: Once the i18n package change is merged, we can add a reference here.

File: packages/patterns/src/isPatternEqual.ts
Comment (swissspidy) [inline packages/patterns/src/isPatternEqual.ts:27]: These are the only valid values AFAIK. Could be moved to `types.ts` as well.

```suggestion
  patternType?: 'color' | 'background'
```

File: packages/patterns/src/generatePatternStyles.ts
Comment (swissspidy) [inline packages/patterns/src/generatePatternStyles.ts]: Should we use the `PatternType` enum here for the cases?

File: packages/patterns/src/generatePatternStyles.ts
Comment (swissspidy) [inline packages/patterns/src/generatePatternStyles.ts]: Wondering if we could define the gradient types in a way that the assertions here are not necessary, i.e. since `pattern.type === 'radial' we _know_ it's a `Radial`.

Comment (swissspidy) [reply packages/patterns/src/generatePatternStyles.ts]: Ah, just saw https://github.com/GoogleForCreators/web-stories-wp/pull/12200#discussion_r960552291

File: packages/patterns/tsconfig.json
Comment (barklund) [reply packages/patterns/tsconfig.json:7]: Should I wait for that package to be merged before merging this? To remove the lint errors?

File: packages/patterns/src/isPatternEqual.ts
Comment (barklund) [reply packages/patterns/src/isPatternEqual.ts:27]: Isn't it used for `border-color` too? But yes, it makes sense to narrow this down as much as possible. I'll look into it.

File: packages/patterns/src/generatePatternStyles.ts
Comment (barklund) [reply packages/patterns/src/generatePatternStyles.ts]: Oh yes, my bad.

File: packages/patterns/src/generatePatternStyles.ts
Comment (barklund) [reply packages/patterns/src/generatePatternStyles.ts]: Here it's actually because `Gradient` isn't a _supertype_ of `Linear` and `Radial`. Maybe I should change the types a bit so this will work. That probably requires interfaces though. I'll check to see if it can be improved.

File: packages/patterns/tsconfig.json
Comment (swissspidy) [reply packages/patterns/tsconfig.json:7]: I just went ahead and merged #12198, so just needs a merge/rebase now!

File: packages/patterns/src/generatePatternStyles.ts
Comment (barklund) [reply packages/patterns/src/generatePatternStyles.ts]: Fixed it in c986cafb77.

File: packages/patterns/src/isPatternEqual.ts
Comment (barklund) [reply packages/patterns/src/isPatternEqual.ts:27]: It is also used for `fill` and `stroke`.

But there are also other cases making narrowing tricky. [Here](https://github.com/GoogleForCreators/web-stories-wp/blob/512328ada20ed77687817ee479bd09255836ca61/packages/element-library/src/product/display.js#L132) it is used for a CSS variable (`--box-shadow-color`), which isn't an unreasonable usage, and the naming is completely arbitrary.

We could enforce a certain number of properties plus any string that starts with `'--'`, but that seems like overkill?

File: packages/patterns/src/generatePatternStyles.ts
Comment (barklund) [reply packages/patterns/src/generatePatternStyles.ts]: Fixed in c986cafb77 (in this and many other places).

**Pull request merged**
